### PR TITLE
chore(typechecking): typecheck /lib

### PR DESCRIPTION
**Pull request by pdesoyres-cc**
## what is this PR

This PR adds typechecking on almost all files in `/lib`.

There are still TS issues when importing modules from clever client.
This prevents us from including all `/lib` directory in `typecheck:ci` task. 

## todo

* [x] merge https://github.com/CleverCloud/clever-components/pull/1120

### CODE REVIEW

File: src/lib/ansi/ansi-palette-analyser.js [hunk 1]
```diff
@@ -23,6 +34,7 @@ export function analyzePalette(palette) {
       compliantCount++;
     }
     ratioSum += ratio;
+    // @ts-ignore
```
Comment (florian-sanders-cc) [inline src/lib/ansi/ansi-palette-analyser.js:37]: **suggestion:** if you type `contrasts` with this, you should be fine

```ts
  /** @type {Partial<Record<ColorsType, { ratio: number, compliant: boolean }>>} */
  const contrasts = {};
```

File: src/lib/ansi/ansi-palette-analyser.js [hunk 2]
```diff
@@ -34,8 +46,14 @@ export function analyzePalette(palette) {
...
+/**
+ *
+ * @param {boolean} isDarkPalette
+ * @return {Array<ColorsType>}
+ */
 function getColorsToTest(isDarkPalette) {
-  const result = [...colors, isDarkPalette ? 'white' : 'black'];
+  const result = [...COLORS, isDarkPalette ? 'white' : 'black'];
 
+  // @ts-ignore
```
Comment (florian-sanders-cc) [inline src/lib/ansi/ansi-palette-analyser.js:57]: **note:** I have the same case within the `cc-ansi-palette` and I haven't found a solution for this either, I'll go with a `ts-expect-error` / ignore as well

File: src/lib/send-to-api.js [hunk 3]
```diff
@@ -13,9 +13,9 @@ import { dispatchCustomEvent } from './events.js';
 
 /**
  * @param {Object} settings
- * @param {ApiConfig} settings.apiConfig
+ * @param {ApiConfig} [settings.apiConfig]
  * @param {AbortSignal} [settings.signal]
- * @param {Number} [settings.cacheDelay]
+ * @param {number} [settings.cacheDelay]
  * @param {Number} [settings.timeout]
```
Comment (florian-sanders-cc) [inline src/lib/send-to-api.js:19]: **nitpick:** 
```suggestion
 * @param {number} [settings.timeout]
```

Comment (florian-sanders-cc) [reply src/lib/send-to-api.js:19]: there are others in the `sendToWarp` function params as well

**Review by florian-sanders-cc — CHANGES_REQUESTED** [review]
Gg @pdesoyres-cc this is great!

I've only added a few nitpicks & some suggestions to avoid a few `@ts-ignore` but apart from that it's all good.

I must admit I did not have enough time to review the smart component part thoroughly so I'll get back to that when I come back if it's not already merged.

**Comment by florian-sanders-cc** [comment]
If we want to benefit from the `typecheck:stat` command, we need to explicitely add all libs to the `include` array:

<details>
  <summary>I've done it in my PR if you want the list</summary>

```json
    "src/lib/",
    "src/lib/utils.js",
    "src/translations",
    "src/lib/animate.js",
    "src/lib/ansi",
    "src/lib/ansi/ansi.js",
    "src/lib/ansi/ansi-palette-analyser.js",
    "src/lib/ansi/ansi-palette-style.js",
    "src/lib/ansi/palettes",
    "src/lib/ansi/palettes/default.js",
    "src/lib/ansi/palettes/everblush.js",
    "src/lib/ansi/palettes/hyoob.js",
    "src/lib/ansi/palettes/night-owl.js",
    "src/lib/ansi/palettes/one-light.js",
    "src/lib/ansi/palettes/tokyo-night-light.js",
    "src/lib/api-helpers.js",
    "src/lib/buffer.js",
    "src/lib/change-case.js",
    "src/lib/clipboard.js",
    "src/lib/color.js",
    "src/lib/css-custom-properties.js",
    "src/lib/date/date-formatter.js",
    "src/lib/date/date-utils.js",
    "src/lib/define-smart-component.js",
    "src/lib/domain.js",
    "src/lib/dom.js",
    "src/lib/events.js",
    "src/lib/fake-strings.js",
    "src/lib/focus-helper.js",
    "src/lib/form",
    "src/lib/form/cc-form-control-element.abstract.js",
    "src/lib/form/form-error-focus-controller.js",
    "src/lib/form/form-submit-directive.js",
    "src/lib/form/form-submit-handler.js",
    "src/lib/form/form-utils.js",
    "src/lib/form/validation.js",
    "src/lib/i18n",
    "src/lib/i18n/i18n-date.js",
    "src/lib/i18n/i18n-display.js",
    "src/lib/i18n/i18n.js",
    "src/lib/i18n/i18n-number.js",
    "src/lib/i18n/i18n-sanitize.js",
    "src/lib/i18n/i18n-string.js",
    "src/lib/i18n/i18n.types.d.ts",
    "src/lib/immer.js",
    "src/lib/leaflet-esm.js",
    "src/lib/leaflet-heat.js",
    "src/lib/memory-cache.js",
    "src/lib/notifications.js",
    "src/lib/pricing.js",
    "src/lib/product.js",
    "src/lib/regex-parse.js",
    "src/lib/remote-assets.js",
    "src/lib/send-to-api.js",
    "src/lib/shadow-dom-utils.js",
    "src/lib/simpleheat.js",
    "src/lib/smart-manager.js",
    "src/lib/smart-symbols.js",
    "src/lib/utils.js",
    "src/lib/xml-parser.js",
    "src/lib/zone.js",

```
</details>

Still, as pointed out within the typecheck:stat PR, we should probably improve the task to avoid having to do that. We could rely on `--listFilesOnly` to do so :thinking:. I thought it could wait but looking at this makes me think that maybe not.

**Review by HeleneAmouzou — APPROVED** [review]
Well done :clap: I did not review the smart component relative files and I focused on the other files. I have no comments to make, but I have discovered unknown ways of typechecking, thanks ! I will come back later when the other  will have review the PR, considering the subject I think it might be interesting for me to read the reviews from the team.

**Review by hsablonniere — APPROVED** [review]
Impressive work @pdesoyres-cc, well done!!!

> There are still TS issues when importing modules from clever client.

It seems like there will be improvements on this soon :tm: 

https://github.com/CleverCloud/clever-client.js/issues/57#issuecomment-2324250363

Really nice that you added function descriptions in the JSDoc!!!

**Comment by florian-sanders-cc** [comment]
> This prevents us from including all /lib directory in typecheck:ci task.
> >       Can we add the files that don't have this problem, one by one?

We could use `// @ts-expect-error FIXME: remove when clever-client exports types` above every clever-client import: https://github.com/CleverCloud/clever-components/pull/1136/files#diff-c825a5f405ed2499499eb16a999c4060e9176b5561c57108c2a5c55ea100b3bdR1

This allows us to typecheck the rest of the file and when the client exposes types, the `@ts-expect-error` will be flagged as useless by the compiler so we'll get errors telling us to remove them :wink:

**Review by florian-sanders-cc — APPROVED** [review]
LGTM! Great work @pdesoyres-cc :sunglasses:

**Pull request merged**